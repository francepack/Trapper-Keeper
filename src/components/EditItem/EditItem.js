import React, { Component } from 'react'
import PropTypes from 'prop-types'

export class EditItem extends Component {
  constructor(props) {
    super(props)
    this.state = {
      value: '',
      id: '',
      completed: false,
    }
  }

  componentDidMount = () => {
    this.setState({
      value: this.props.value,
      id: this.props.id,
      completed: this.props.completed,
    })
  }

  updateItem = (e) => {
    this.setState({
      value: e.target.value
    }, () => {this.props.updateItem(this.state)})
  }

  checkKey = (e) => {
    if(e.keyCode === 13) {
      e.preventDefault()
    }
  }

  deleteItem = (e) => {
    e.preventDefault()
    this.setState({value: ''})
    this.props.delete(this.state.id)
  }

  toggleCompleted = () => {
    let complete = !this.state.completed
    const {value, id} = this.state
    this.setState({completed: complete})
    this.props.moveCompleted({value, id, completed: complete})
  }

  render() {
    return (
      <div className='list-container' >
          <input type='checkbox'
                  className='list-control complete-item'
                  name='completed'
                  checked={this.state.completed}
                  onChange={this.toggleCompleted}
                  >
          </input>
        {
          this.state.id &&
          <textarea onKeyUp={this.updateItem}
                    onKeyDown={this.checkKey}
                    className='list-item'
                    defaultValue={this.state.value}
                    placeholder='Add an item . . .'
                    >
          </textarea>
        }
        <button className='list-control delete' onClick={this.deleteItem}><i class="fas fa-minus-circle delete-item"></i></button>
      </div>
    )
  }
}

EditItem.propTypes = {
  updateNoteItems: PropTypes.func,
  delete: PropTypes.func,
  item: PropTypes.object
}

export default EditItem