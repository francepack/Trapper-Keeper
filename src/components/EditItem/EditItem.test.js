import EditItem from './EditItem'
import React from 'react'
import { shallow, mount } from 'enzyme'

describe('EditItem', () => {
  let wrapper

  beforeEach(() => {
    const props = {
      handleChange: jest.fn(),
      handleItem: jest.fn(),
      updateNoteItems: jest.fn(),
      delete: jest.fn(),
      value: 'test',
      id: '123',
      completed: false
    }

    Date.now = jest.fn().mockImplementation(() => 6)
    wrapper = shallow(
      <EditItem {...props}/>
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have a default state set on mount to match props', () => {
    expect(wrapper.state()).toEqual( {value: 'test', id: '123', completed: false} )
  })

  describe('update item', () => {
    it('should update value state onChange and call handleChange', () => {
      const mockState = {
        value: 'he',
        id: 6,
        completed: false,
      }
  
      const mockEvent = {
        target: {
          value: 'hey'
        }
      }
  
      wrapper.setState(mockState)
      expect(wrapper.instance().state).toEqual(mockState)
  
      wrapper.instance().updateItem(mockEvent)
      expect(wrapper.state('value')).toEqual('hey')
    })
  })

  describe('editItem', () => {
    it('should be called when a the correct button is clicked', () => {
      const e = Object.assign(jest.fn(), {preventDefault: () => {}})
      
      wrapper.instance().editItem = jest.fn()
      wrapper.find('.confirm-item').simulate('click', e)
      expect(wrapper.instance().editItem).toBeCalled()
    })

    it('should call updateNoteItems if there is value in the input', () => {
      const mockState = {
        value: 'he',
        id: 6,
        completed: false,
      }
      const mockEmptyState = {
        value: '',
        id: 6,
        completed: false,
      }
      const e = Object.assign(jest.fn(), {preventDefault: () => {}})
      
      wrapper.setState(mockEmptyState)
      wrapper.instance().editItem(e)
      expect(wrapper.instance().props.updateNoteItems).not.toBeCalled()
      
      wrapper.setState(mockState)
      wrapper.instance().editItem(e)
      expect(wrapper.instance().props.updateNoteItems).toBeCalled()
    })
  })

  describe('deleteItem', () => {
    it('should be called when the correct button is clicked', () => {
      const e = Object.assign(jest.fn(), {preventDefault: () => {}})
      wrapper.find('.delete-item').simulate('click', e)
      expect(wrapper.instance().deleteItem).toBeCalled()
    })

    it('should clear state value', () => {
      
    })

    it('should call delete with the correct id', () => {
      
    })
  })

  describe('toggleCompleted', () => {
    it('should be called when the checkbox is changed', () => {
      wrapper.instance().toggleCompleted = jest.fn()
      wrapper.find('.complete-item').simulate('change')
      expect(wrapper.instance().toggleCompleted).toBeCalled()
    })

    it('should toggle state complete', () => {
      const mockState = {
        value: 'he',
        id: 6,
        completed: false,
      }
      const expected = {
        value: 'he',
        id: 6,
        completed: true,
      }
      wrapper.setState(mockState)
      wrapper.instance().toggleCompleted()
      expect(wrapper.state()).toEqual(expected)
    })
  })
})

  // describe('handleItemBlur', () => {
  //   it('should call handleItem with the properties in state', () => {
  //     const mockState = {
  //       value: 'he',
  //       id: 6,
  //       completed: false,
  //     }
  //     wrapper.setState(mockState)
  //     expect(wrapper.instance().state).toEqual(mockState)
  
  //     wrapper.instance().handleItemBlur()
  //     expect(wrapper.instance().props.handleItem).toHaveBeenCalledWith(mockState)
  //   })
  
  //   it('should not call handleItem if the value in state is and empty string', () => {
  //     const mockState = {
  //       value: '',
  //       id: 6,
  //       completed: false,
  //     }
  //     wrapper.setState(mockState)
  //     expect(wrapper.instance().state).toEqual(mockState)
  
  //     wrapper.instance().handleItemBlur()
  //     expect(wrapper.instance().props.handleItem).not.toHaveBeenCalled()
  //   })
  // })