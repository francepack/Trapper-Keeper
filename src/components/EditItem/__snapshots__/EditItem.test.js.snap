// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditItem should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditItem
    completed={false}
    delete={[MockFunction]}
    id="123"
    moveCompleted={[MockFunction]}
    updateItem={[MockFunction]}
    value="test"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={false}
          className="list-control complete-item"
          name="completed"
          onChange={[Function]}
          type="checkbox"
        />,
        <textarea
          className="list-item"
          defaultValue="test"
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          placeholder="Add an item . . ."
        />,
        <button
          className="list-control delete"
          onClick={[Function]}
        >
          <i
            class="fas fa-minus-circle delete-item"
          />
        </button>,
      ],
      "className": "list-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": false,
          "className": "list-control complete-item",
          "name": "completed",
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "list-item",
          "defaultValue": "test",
          "onKeyDown": [Function],
          "onKeyUp": [Function],
          "placeholder": "Add an item . . .",
        },
        "ref": null,
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            class="fas fa-minus-circle delete-item"
          />,
          "className": "list-control delete",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "class": "fas fa-minus-circle delete-item",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={false}
            className="list-control complete-item"
            name="completed"
            onChange={[Function]}
            type="checkbox"
          />,
          <textarea
            className="list-item"
            defaultValue="test"
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            placeholder="Add an item . . ."
          />,
          <button
            className="list-control delete"
            onClick={[Function]}
          >
            <i
              class="fas fa-minus-circle delete-item"
            />
          </button>,
        ],
        "className": "list-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": false,
            "className": "list-control complete-item",
            "name": "completed",
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "list-item",
            "defaultValue": "test",
            "onKeyDown": [Function],
            "onKeyUp": [Function],
            "placeholder": "Add an item . . .",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              class="fas fa-minus-circle delete-item"
            />,
            "className": "list-control delete",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "class": "fas fa-minus-circle delete-item",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
